### PR TITLE
[idrac_gather_facts][role]: Enhancements - Controller & SecureBoot

### PR DESCRIPTION
**Pull request by jagadeeshnv**
- iDRAC: Add storage controller details in the idrac_gather_facts role output
- iDRAC - Secure Boot support

### CODE REVIEW

**Comment by ansible-collections-svc** [comment]
Can one of the admins verify this patch?

File: roles/idrac_gather_facts/tasks/main.yml [hunk 1]
```diff
@@ -1,162 +1,103 @@
...
+      method: "{{ idrac_gather_facts_uri_method }}"
+      user: "{{ username | default(lookup('env', 'IDRAC_USERNAME')) }}"
+      password: "{{ password | default(lookup('env', 'IDRAC_PASSWORD')) }}"
+      timeout: "{{ https_timeout }}"
+      force_basic_auth: true
+      headers: "{{ idrac_gather_facts_uri_headers }}"
+      body_format: "{{ idrac_gather_facts_uri_body_format }}"
+      status_code: "{{ idrac_gather_facts_uri_status_code }}"
+      return_content: "{{ idrac_gather_facts_uri_return_content }}"
+  vars:
```
Comment (sachin-apa) [inline roles/idrac_gather_facts/tasks/main.yml]: couldn't we move this to vars.yml ?

Comment (jagadeeshnv) [reply roles/idrac_gather_facts/tasks/main.yml]: module_defaults is better, as will apply to uri module only

File: roles/idrac_gather_facts/tasks/main.yml [hunk 2]
```diff
@@ -1,162 +1,79 @@
...
-
-- name: Gather PhysicalDisk facts.
-  ansible.builtin.include_tasks: get_physical_info.yml
-  when: target is defined and "PhysicalDisk" in target
-
-- name: Gather SystemMetrics facts.
-  ansible.builtin.include_tasks: get_metrics_info.yml
-  when: target is defined and "SystemMetrics" in target
+- name: IDRAC Gather facts
+  module_defaults:
```
Comment (ABHISHEK-SINHA10) [inline roles/idrac_gather_facts/tasks/main.yml:3]: Nice 👍 It'll increase code readability and unnecessary duplicate line.

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml [hunk 3]
```diff
@@ -10,30 +11,67 @@
...
+  register: controllers_url
+
+- name: Pick Controllers key value in one list
   ansible.builtin.set_fact:
-    diff_keys: "{{ controller_data.keys() | list | symmetric_difference((api_response.keys() | list)) }}"
+    controllers_list: "{{ controllers_list | default([]) + [item.json.Controllers] }}"
+  loop: "{{ controllers_url.results }}"
 
-- name: Set a diff of dict
+- name: Select list of values from dictionary
```
Comment (ABHISHEK-SINHA10) [inline roles/idrac_gather_facts/tests/asserts/controller_assert.yml]: Please mention which dictionary in name, seems generic.

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml [hunk 4]
```diff
@@ -10,30 +11,67 @@
...
-- name: Set controller facts
+- name: Get storage entities
   ansible.builtin.set_fact:
-    api_response: "{{ controller_result.json | ansible.utils.remove_keys(target=['@odata.context', '@odata.id', '@odata.type']) }}"
+    collected_storage_entity: "{{ (collected_storage_entity | default([])) + [item['@odata.id']] }}"
+  loop: "{{ storage_entity.json.Members }}"
 
-- name: Set the keys diff
+- name: Get Controllers URL information.
+  ansible.builtin.uri:
```
Comment (ABHISHEK-SINHA10) [inline roles/idrac_gather_facts/tests/asserts/controller_assert.yml:24]: can't we have same module_defaults here?

File: roles/idrac_gather_facts/README.md [hunk 5]
```diff
@@ -298,6 +667,7 @@ ansible.utils
       - PCIeSSDBackPlane
       - PowerSupply
       - PresenceAndStatusSensor
+      - SecureBoot
```
Comment (Jennifer-John) [inline roles/idrac_gather_facts/README.md:713]: can we update the examples to include displaying the output? (as in bspec)
Eg:
ansible.builtin.import_role:
    name: idrac_gather_facts
vars:
    << creds >>
    target:
      - System
      - Chassis
      - Manager
  tags:
    - idrac_gather_facts_out
 
-name: Print the System details
  ansible.builtin.debug:
    var: system
  tags:
    -idrac_gather_facts_out

File: roles/idrac_gather_facts/README.md [hunk 6]
```diff
@@ -119,7 +118,7 @@ ansible.utils
     <td>- System <br> - BIOS <br> - Controller <br> - CPU <br> - Enclosure <br> - EnclosureEMM <br> - Fan <br>
         - Firmware <br> - HostNIC <br> - License <br> - Memory <br> - NIC <br> - PCIeSSDBackPlane <br>
         - PowerSupply <br> - PresenceAndStatusSensor <br> - Sensors_Battery <br> - Sensors_Intrusion <br>
-        - Sensors_Voltage <br> - VirtualDisk <br> - PCIeDevice <br> - PhysicalDisk <br> - SystemMetrics</td>
+        - Sensors_Voltage <br> - VirtualDisk <br> - PCIeDevice <br> - PhysicalDisk <br> - SystemMetrics<br> - SecureBoot</td>
     <td>list</td>
     <td>- I(target) component for which information needs to be gathered.</td>
```
Comment (Jennifer-John) [inline roles/idrac_gather_facts/README.md]: can we update the component specific description as well here? (as in the bspecs)

File: roles/idrac_gather_facts/tasks/get_controller_info.yml [hunk 7]
```diff
@@ -1,21 +1,32 @@
...
 
-- name: Set Controller facts
+- name: Get all storage controllers.
   ansible.builtin.set_fact:
-    controller: "{{ controller_result.json.Members | ansible.utils.remove_keys(target=['@odata.context', '@odata.id', '@odata.type']) }}"
+    controllers_list: "{{ disk_result.json.Members | selectattr('Controllers', 'defined') | map(attribute='Controllers') }}"
+
+- name: Select list of values from dictionary
+  ansible.builtin.set_fact:
+    collected_values: "{{ (collected_values | default([])) + [controller_item['@odata.id']] }}"
```
Comment (Jennifer-John) [inline roles/idrac_gather_facts/tasks/get_controller_info.yml]: can we update the var name to odata_id_list or something similar?

File: roles/idrac_gather_facts/README.md (hunk 6 above)
Comment (jagadeeshnv) [reply roles/idrac_gather_facts/README.md]: Sure

File: roles/idrac_gather_facts/tasks/get_controller_info.yml (hunk 7 above)
Comment (jagadeeshnv) [reply roles/idrac_gather_facts/tasks/get_controller_info.yml]: yes will rename to controller_list_temp

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml (hunk 3 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/controller_assert.yml]: Done

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml (hunk 4 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/controller_assert.yml:24]: Used referenced variables instead.

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml [hunk 8]
```diff
@@ -10,30 +11,60 @@
     body_format: "{{ idrac_gather_facts_uri_body_format }}"
     status_code: "{{ idrac_gather_facts_uri_status_code }}"
     return_content: "{{ idrac_gather_facts_uri_return_content }}"
-  register: controller_result
+  register: storage_entity
```
Comment (sachin-apa) [inline roles/idrac_gather_facts/tests/asserts/controller_assert.yml]: can you rename this register var to idrac_gather_facts_storage_entity

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml [hunk 9]
```diff
@@ -10,30 +11,60 @@
...
+    collected_storage_entity:
+      "{{ (collected_storage_entity | default([])) + [item['@odata.id']] }}"
+  loop: "{{ storage_entity.json.Members }}"
+
+- name: Get Controllers URL information.
+  ansible.builtin.uri:
+    url: "https://{{ hostname }}{{ item }}"
+    <<: *uri_params
+  loop: "{{ collected_storage_entity }}"
+  register: controllers_url
```
Comment (sachin-apa) [inline roles/idrac_gather_facts/tests/asserts/controller_assert.yml]: can you rename this register var to idrac_gather_facts_

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml [hunk 10]
```diff
@@ -10,30 +11,60 @@
...
+       [item['@odata.id'] | default(omit)] }}"
+  loop: "{{ controllers_list }}"
 
-- name: Set a diff of dict
+- name: Get Controllers information.
+  ansible.builtin.uri:
+    url: "https://{{ hostname }}{{ item }}?$expand=*($levels=1)"
+    <<: *uri_params
+  loop: "{{ collected_values | reject('match', '__omit_place_holder__.*') }}"
+  register: result
```
Comment (sachin-apa) [inline roles/idrac_gather_facts/tests/asserts/controller_assert.yml]: can you rename this register var to idrac_gather_facts_

File: roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    validate_certs: "{{ validate_certs }}"
+    ca_path: "{{ ca_path | default(omit) }}"
+    method: "{{ idrac_gather_facts_uri_method }}"
+    user: "{{ username }}"
+    password: "{{ password }}"
+    headers: "{{ idrac_gather_facts_uri_headers }}"
+    body_format: "{{ idrac_gather_facts_uri_body_format }}"
+    status_code: "{{ idrac_gather_facts_uri_status_code }}"
+    return_content: "{{ idrac_gather_facts_uri_return_content }}"
+  register: secureboot_details
```
Comment (sachin-apa) [inline roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml]: can you rename this register var to idrac_gather_facts_

File: roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+- name: Fetching Secureboot database info
+  ansible.builtin.uri:
+    url: "https://{{ hostname }}{{ api_system
+      }}/SecureBoot/SecureBootDatabases?$expand=*($levels=1)"
+    <<: *uri_params
+  register: secureboot_database_details
```
Comment (sachin-apa) [inline roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml]: can you rename this register var to idrac_gather_facts_

File: roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+- name: Fetching Secureboot database certificates info
+  ansible.builtin.uri:
+    url: "https://{{ hostname }}{{
+      sec_boot_db_item['Certificates']
+      ['@odata.id'] }}?$expand=*($levels=1)"
+    <<: *uri_params
+  loop: "{{ secureboot_database_details.json.Members }}"
+  loop_control:
+    loop_var: sec_boot_db_item
+  register: secure_boot_results
```
Comment (sachin-apa) [inline roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml]: can you rename this register var to idrac_gather_facts_

File: roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+- name: Check whether output differs
+  ansible.builtin.set_fact:
+    result_diff: "{{ secure_boot | symmetric_difference(api_response) }}"
+
+- name: Assert the differences in List
+  ansible.builtin.assert:
+    that:
+      - "{{ (result_diff | length) == 0 }}"
+    fail_msg: "The response from the role does not match"
```
Comment (sachin-apa) [inline roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml:59]: possible to declare this above. line 61 and 62

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml [hunk 15]
```diff
@@ -10,30 +11,60 @@
...
     status_code: "{{ idrac_gather_facts_uri_status_code }}"
     return_content: "{{ idrac_gather_facts_uri_return_content }}"
-  register: controller_result
+  register: storage_entity
   no_log: true
 
-- name: Set controller facts
+- name: Get storage entities
+  ansible.builtin.set_fact:
+    collected_storage_entity:
```
Comment (sachin-apa) [inline roles/idrac_gather_facts/tests/asserts/controller_assert.yml:19]: can you rename this register var to idrac_gather_facts_

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml (hunk 8 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/controller_assert.yml]: Done

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml (hunk 9 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/controller_assert.yml]: Done

File: roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml (hunk 11 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml]: Done

File: roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml (hunk 13 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml]: Done

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml (hunk 15 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/controller_assert.yml:19]: Done

File: roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml (hunk 14 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml:59]: Done

File: roles/idrac_gather_facts/tests/asserts/controller_assert.yml (hunk 10 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/controller_assert.yml]: Done

File: roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml (hunk 12 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/tests/asserts/secureboot_assert.yml]: Done

File: roles/idrac_gather_facts/molecule/secureboot/converge.yml [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+    - name: Assert secureboot supply dict for length
+      ansible.builtin.assert:
+        that:
+          - "{{ secure_boot | length > 0 }}"
+
+    - name: Call assertion
+      ansible.builtin.include_tasks: ../../tests/asserts/secureboot_assert.yml
+      # with_items: "{{ secure_boot }}"
+      # loop_control:
```
Comment (ABHISHEK-SINHA10) [inline roles/idrac_gather_facts/molecule/secureboot/converge.yml]: Remove these commented line

File: roles/idrac_gather_facts/tasks/get_secure_boot_info.yml [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+      }}/SecureBoot/?$expand=*($levels=1)
+  register: secure_boot_result
+  delegate_to: "{{ idrac_gather_facts_delegate }}"
+- name: Get Secure Boot Databases.
+  ansible.builtin.uri:
+    url: https://{{ hostname }}:{{ https_port }}{{ api_system
+      }}/SecureBoot/SecureBootDatabases?$expand=*($levels=1)
+  register: secure_boot_db_result
+  delegate_to: "{{ idrac_gather_facts_delegate }}"
+- name: Get Secureboot db certificates information.
```
Comment (ABHISHEK-SINHA10) [inline roles/idrac_gather_facts/tasks/get_secure_boot_info.yml:16]: Give space between tasks, it's not readable

File: roles/idrac_gather_facts/molecule/secureboot/converge.yml (hunk 16 above)
Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/molecule/secureboot/converge.yml]: Done

**Review by sachin-apa — DISMISSED** [review]

**Review by Kritika-Bhateja-03 — DISMISSED** [review]

File: roles/idrac_gather_facts/molecule/secureboot/converge.yml [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+      Accept: "application/json"
+      Content-Type: "application/json"
+      OData-Version: "4.0"
+    idrac_gather_facts_uri_body_format: "json"
+    idrac_gather_facts_uri_status_code:
+      - 200
+      - 400
+      - 401
+      - 404
+      - -1
```
Comment (Jennifer-John) [inline roles/idrac_gather_facts/molecule/secureboot/converge.yml:23]: just checking if -1 is also a valid status code for the idrac api?

Comment (Kritika-Bhateja-03) [reply roles/idrac_gather_facts/molecule/secureboot/converge.yml:23]: Yes, it might occur if connection was aborted before a complete request.

**Review by sachin-apa — APPROVED** [review]

**Review by Jennifer-John — APPROVED** [review]

**Pull request merged**